### PR TITLE
[Enhancement] enable simple count using meta scan

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

For full table count aggregation query without filter, we can use meta scan to get result from segment metadata instead of reading a large amount of data.
In this PR, i extend RewriteMinMaxAggToMetaScanRule to support such rewriting and rename this rule to RewriteSimpleAggToMetaScanRule. whether enable this rule is determined by session variable `enable_rewrite_simple_agg_to_meta_scan`.

**NOTE**
Before this change, the columns returned by meta scan are not null. When querying an empty table, `count` will return NULL instead of 0, which is wrong. In order to fix this problem, I changed the returned column to nullable, and return a row of mock data to ensure the correct result when querying an empty table. In the mock result, count returns 0, and other columns return NULL.

Currently, meta scan will only be used in the background task of collecting low-cardinality dictionaries in FE. Changing the nullable attribute of a column will cause incompatibility during upgrade. To avoid this problem, we need to determine whether the aggregation function includes `count` or not. if `count` is not included, then all columns returned by meta scan are not null, the behavior is the same as before, otherwise they will be changed to nullable.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 66 / 66 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteSimpleAggToMetaScanRule.java|56|56|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java|3|3|100.00%|[]|

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 66 / 66 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteSimpleAggToMetaScanRule.java|56|56|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java|3|3|100.00%|[]|

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run all

**Comment by silverbullet233** [comment]
run all

**Comment by silverbullet233** [comment]
run all

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 16 / 66 (24.24%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|0|4|00.00%|[1781, 1782, 1784, 1785]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteSimpleAggToMetaScanRule.java|10|56|17.86%|[66, 70, 71, 72, 73, 74, 76, 83, 84, 85, 87, 88, 91, 95, 97, 98, 102, 103, 104, 105, 107, 108, 110, 172, 173, 175, 176, 177, 178, 180, 183, 184, 186, 187, 188, 189, 190, 192, 194, 195, 196, 197, 199, 200, 202, 216]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 16 / 66 (24.24%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|0|4|00.00%|[1781, 1782, 1784, 1785]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteSimpleAggToMetaScanRule.java|10|56|17.86%|[66, 70, 71, 72, 73, 74, 76, 83, 84, 85, 87, 88, 91, 95, 97, 98, 102, 103, 104, 105, 107, 108, 110, 172, 173, 175, 176, 177, 178, 180, 183, 184, 186, 187, 188, 189, 190, 192, 194, 195, 196, 197, 199, 200, 202, 216]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java|3|3|100.00%|[]|

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 60 / 67 (89.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteSimpleAggToMetaScanRule.java|50|57|87.72%|[175, 197, 198, 199, 201, 202, 204]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java|3|3|100.00%|[]|

**Comment by silverbullet233** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java [hunk 1]
```diff
@@ -126,7 +126,10 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
             }
 
             aggColumnIdToNames.put(metaColumn.getId(), metaColumnName);
-            newScanColumnRefs.put(metaColumn, metaScan.getColRefToColumnMetaMap().get(usedColumn));
+            Column c = metaScan.getColRefToColumnMetaMap().get(usedColumn);
+            // for empty table, meta scan may return NULL result, so we force set usedColumn to nullable.
+            c.setIsAllowNull(true);
+            newScanColumnRefs.put(metaColumn, c);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java:132]: soo, modify column-ref can't work?

Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java:132]: > modify column-ref can't work?

not work, seems nullable property depens on Column::isAllowNull, not column ref

File: be/src/storage/meta_reader.cpp [hunk 2]
```diff
@@ -90,7 +113,7 @@ Status MetaReader::_fill_result_chunk(Chunk* chunk) {
             TypeDescriptor desc;
             desc.type = TYPE_ARRAY;
             desc.children.emplace_back(item_desc);
-            ColumnPtr column = ColumnHelper::create_column(desc, false);
+            ColumnPtr column = ColumnHelper::create_column(desc, true);
```
Comment (stdpain) [inline be/src/storage/meta_reader.cpp:116]: I'm not sure whether if works when grayscale upgrade

**Comment by silverbullet233** [comment]
run starrocks_admit_test

File: be/src/storage/meta_reader.cpp (hunk 2 above)
Comment (silverbullet233) [reply be/src/storage/meta_reader.cpp:116]: > I'm not sure whether if works when grayscale upgrade

meta scan is only used when collecting low-cardinality dict, can we disable it when upgrading? @stdpain

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 57 / 66 (86.36%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[1882, 1883]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteSimpleAggToMetaScanRule.java|48|55|87.27%|[171, 193, 194, 195, 197, 198, 200]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java|3|3|100.00%|[]|

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

File: be/src/storage/meta_reader.cpp [hunk 3]
```diff
@@ -199,6 +231,10 @@ Status SegmentMetaCollecter::_collect_dict(ColumnId cid, Column* column, Logical
     auto dst = array_column->elements_column().get();
     CHECK(dst->append_strings(words));
 
+    if (column->is_nullable()) {
+        nullable_column->null_column_data().emplace_back(0);
```
Comment (stdpain) [inline be/src/storage/meta_reader.cpp:235]: miss update_has_null ?

**Review by Seaven — APPROVED** [review]

File: be/src/storage/meta_reader.cpp (hunk 3 above)
Comment (silverbullet233) [reply be/src/storage/meta_reader.cpp:235]: > miss update_has_null ?

no need here, because 0 indicates not null, it won't change the result of has_null

**Review by stdpain — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**